Thank you for this!

I did a review locally and I'm getting `AssertionError` -- it seems like maybe some of the test fixture data you intended is missing from this PR?

Here's how the `tree test/simple/data` looks on my checkout of this PR:

```
❯ tree test/data/simple
test/data/simple
├── BYTE_MD5SUMS
├── dest
│   ├── hello.txt
│   └── sub
│       └── shorter.txt
└── src
    ├── one
    │   └── hello.txt
    └── two
        └── a-much-longer-filename.txt

5 directories, 5 files
```

Here's an example of the `AssertionError` issues I get on running `make test`:

```
        assert not metadata_only, metadata_only
>       assert not filesystem_only, filesystem_only
E       AssertionError: {'./src/two/more/missing.txt'}

test/data.py:95: AssertionError
============================================= short test summary info =============================================
FAILED test/unit/test_bytes.py::test_unknown - AssertionError: {'./src/two/more/missing.txt'}
FAILED test/unit/test_hash.py::test_data - AssertionError: {'./src/two/more/missing.txt'}
FAILED test/unit/test_hash.py::test_simple_hash - AssertionError: {'./src/two/more/missing.txt'}
FAILED test/unit/test_hash.py::test_simple_hashes - AssertionError: {'./src/two/more/missing.txt'}
FAILED test/unit/test_primary_key.py::test_consistency - AssertionError: {'./src/two/more/missing.txt'}
================================================ 5 failed in 0.25s ================================================
make: *** [Makefile:17: test] Error 1
```

This perhaps suggests one of the missing files is:

```
test/data/simple/src/two/more/missing.txt
```

Alternatively, it looks like if I remove the relevant line for this file from `test/data/simple/BYTE_MD5SUMS`, the tests pass.

--- a/test/data/simple/BYTE_MD5SUMS
+++ b/test/data/simple/BYTE_MD5SUMS
@@ -2,4 +2,4 @@
 3f52d078e6dbaa62fe5c1fe5dceafa5d  ./dest/sub/shorter.txt
 746308829575e17c3331bbcb00c0898b  ./src/one/hello.txt
 3f52d078e6dbaa62fe5c1fe5dceafa5d  ./src/two/a-much-longer-filename.txt
-2ff66bc28e8ac8f8a6c14a41bec9a2d1  ./src/two/more/missing.txt
+2ff66bc28e8ac8f8a6c14a41bec9a2d1  ./src/two/more/src-only.txt

--- a/test/data/simple/src/two/more/src-only.txt
+++ b/test/data/simple/src/two/more/src-only.txt
@@ -1,0 +1,1 @@
+Keep this out of destination in order to test missing source.

